compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2DToImg2Col.cpp

```diff
@@ -375,6 +375,155 @@ class ConvertDepthwiseConv2DNhwcHwc final
   }
 };
 
+// For nchw, because the channels are to the left of the image shape dimensions,
+// the position of the contraction dimension in the resulting matmul is
+// reversed. This swaps the LHS and RHS of the matmul when compared with nhwc
+// (i.e. (D, C x Kh x Kw) * (C x Kh x Kw, Ho x Wo))
+class ConvertConv2DNchwFchw final
+    : public OpRewritePattern<linalg::Conv2DNchwFchwOp> {
+ public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNchwFchwOp convOp,
+                                PatternRewriter &rewriter) const override {
+    auto inputType = convOp.getInputs()[0].getType().cast<ShapedType>();
+    auto filterType = convOp.getInputs()[1].getType().cast<ShapedType>();
+    auto outputType = convOp.getOutputs()[0].getType().cast<ShapedType>();
+
+    if (!filterType.hasStaticShape() || !inputType.hasStaticShape()) {
+      return failure();
+    }
+
+    // TODO: Support dilation.
+    if (!hasAllOneValues(convOp.getDilations())) return failure();
+
+    Value input = convOp.getInputs()[0];
+    Value filter = convOp.getInputs()[1];
+    Value output = convOp.getOutputs()[0];
+
+    auto filterShape = filterType.getShape();
+    auto outputShape = outputType.getShape();
+
+    const int n = outputShape[0];
+    const int oc = outputShape[1];
+    const int oh = outputShape[2];
+    const int ow = outputShape[3];
+    const int ic = filterShape[1];
+    const int fh = filterShape[2];
+    const int fw = filterShape[3];
+
+    auto loc = convOp.getLoc();
+
+    SmallVector<int64_t, 4> colTensorShape = {n, ic, fh, fw, oh, ow};
+
+    Value colTensor = rewriter.create<tensor::EmptyOp>(
+        loc, colTensorShape, inputType.getElementType());
+
+    AffineExpr nDim, icDim, khDim, kwDim, ohDim, owDim;
+    bindDims(getContext(), nDim, icDim, khDim, kwDim, ohDim, owDim);
+
+    auto shSym = rewriter.getAffineConstantExpr(
+        convOp.getStrides().getValues<int64_t>()[0]);
+    auto swSym = rewriter.getAffineConstantExpr(
+        convOp.getStrides().getValues<int64_t>()[1]);
+
+    SmallVector<AffineExpr, 4> inputExprs = {nDim, icDim, ohDim * shSym + khDim,
```

Comment:
I see this is how the others are done, but such indexing maps are generally a pain to handle.... Can you write this as broadcast + transpose + reshape of the inputs instead. I think that should achieve what you are looking for.

Reply:
Had an offline discussion about this as well. This might be better represented as a gather... This will be a follow up too.